### DISCUSSION TITLE
how to move forward simulation using callback from C function

### DISCUSSION THREAD

**Discussion by kirknodeng** [cocotb/cocotb#4174] [Q&A]
cocotb version : 1.9.1 
operating system : el8 Linux 
simulator : vcs P2019.06-sp2 
python version : 3.8.12 self-installed 

I'm trying to migrate C function which is system level application to cocotb environment. 
I used "ctypes.CDLL" to call c shared library and run c function in cocotb when after the design initialization. 
First of all I was thinking to interface with design and C function using DPI. but, I realized c function and simulation side scope should be different. (I'm not sure .. it might be somewhere in the library but , it could be not matched in simulator view) 

Anyway, so I changed my mind that use callback from C side. 
As a result, I can call the python callback function from C side. but, there are some problem.. it's not able to go further. it's not async statement. 

RuntimeWarning : coroutine 'test_env.callback' was never awaited. 

Also, I tried to used another way that used "asyncio" also.. but, it's not works. 
Do you have any idea to advance simulation time using callback from C side? 

Thank you

**Comment by kirknodeng** [comment]
What I tried .. In terms of C function call from python side using ctype.CDLL , it had an error. 
but, when set the scope it can fix scope issue. like below 

svSetScope(svGetScopeFromName("tb_wrapper")); 

this statement can make the function call from python side. After that I'd like to advance simulation time from c side function call. 
I was using "export" dpi. but, it is not work. that because it is started from python through C function to systemverilog export call. ( invoke error, related with time consume )
I think in this case simulator can control a schedule simulation time..  

So, I checked my assumption. I used import dpi and export dpi without python side C function calling. 
actually, it is still based on cocotb environment. but, not used c function call in python side. C function call(import) from testbench (systemverilog) side. and C function call systemverilog task. it works. 

Conclusion, it is not able to make a time consume export DPI call ( C side) from pyhton side. like below 
e.g ) 
1. python (cocotb) : 
 lib = ctypes.CDLL("mylib.so") 
 lib.c_function()  # C function call from python side 
2. C ( normal function library )
 extern void sv_dpi_export(void);
 void c_function(void){
    sv_dpi_export();   # export dpi call
 }
3. testbench wrapper ( systemverilog)
  initial begin
  .... 
  end 
  reg a=0;
  export "DPI-C" task sv_dpi_export;
  task sv_dpi_export()
  \# 100 
  a=1;
  endtask

but, I mixed two thing to execute c_function. and cocotb still control simulation. like below 
.g ) most thing is same but I make a flag
1. python (cocotb) : 
  \# didn't use ctype , make a register inside of systemverilog and control
  dut.run_c_function.value = 1
  await FallingEdge(dut.run_c_function)

2. C ( normal function library )
 extern void sv_dpi_export(void);
 void c_function(void){
    sv_dpi_export();   # export dpi call
 }
3. testbench wrapper ( systemverilog)
  initial begin
  .... 
  end 
  reg a=0;
  export "DPI-C" task sv_dpi_export;
  task sv_dpi_export()
  \# 100 
  a=1;
  endtask

  reg run_c_function = 0;
 always @(posedge run_c_function) begin
  if (run_c_function) begin
     c_function();
  end

import "DPI" context void c_function(void);  
end 

Some of the syntax could be made an error.. I couldn't paste original one.. 
It would be happy that could be helpful to make this kind of functionality.. 
and, also.. I am still curios about callback function.. it could be make the time consume functionality.. between c, python and systemverilog... I'm not sure..

**Comment by kirknodeng** [comment]
Is there any concern, question, recommendation, it would be helpful to me.

**Comment by ktbarrett** [comment]
I would check out https://github.com/fvutils/pyhdl-if. cocotb was not set up to interact with the DPI. The VPI (which cocotb uses) and DPI are fundamentally different types of interfaces and somewhat incompatible. pyhdl-if was designed to provide functionality to bridge the two.
